### PR TITLE
Add WASI Implementation

### PR DESCRIPTION
**Pull request by satabin**
This PR adds first implementation of the WASI API. It is based on the preview snapshot 1 and focuses on the file operations.
Since sockets will be changed a lot in the next snapshot, it is not implemented here (and there is no good way to use them for now in WASI as no socket can be open).

[asciicast](https://asciinema.org/a/FS1ApqL5zAh301inueIWOXJg3)

### CODE REVIEW

File: wasi/core/src/swam/runtime/wasi/Core.scala [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+        F.pure(inval)
+    }
+
+  def fdAdvise(fd: Fd, offset: Filesize, len: Filesize, advice: Advice): F[Errno] =
+    unimplemented("fd_advise")
+
+  def fdAllocate(fd: Fd, offset: Filesize, len: Filesize): F[Errno] =
+    unimplemented("fs_allocate")
+
+  def pathOpen(fd: Fd,
```
Comment (satabin) [inline wasi/core/src/swam/runtime/wasi/Core.scala:234]: @ybasket Just as a sanity check, could you please have a look at my use of cats-effect in this method, and more generally this class? :pray:

File: build.sc [hunk 2]
```diff
@@ -27,40 +27,43 @@ val swamUrl = "https://github.com/satabin/swam"
...
 
   def scalacOptions =
-    Seq("-feature", "-deprecation", "-unchecked", "-Ypatmat-exhaust-depth", "off", "-Ywarn-unused:locals,imports", "-Ymacro-annotations")
+    Seq("-feature",
+        "-deprecation",
+        "-unchecked",
+        "-Ypatmat-exhaust-depth",
+        "off",
+        "-Ywarn-unused:locals,imports",
+        "-Ymacro-annotations")
```
Comment (ybasket) [inline build.sc:45]: You might want to add `-Wvalue-discard`, it should have caught one of the things below

File: wasi/core/src/swam/runtime/wasi/Core.scala [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        }
+
+    (for {
+      _ <- mem.writeInt(environc, count)
+      _ <- mem.writeInt(environBufSize, size)
+    } yield success)
+      .handleErrorWith(t => logger.error("Could not get the environment size", t).as(io))
+  }
+
+  // TODO I assume nanosecond precision is available but it depends on the `Clock` implementation
```
[the remark this answers is not in the document]
Comment (satabin) [reply wasi/core/src/swam/runtime/wasi/Core.scala:192]: Yeah it's more of a note, than a TODO, I will change that.

File: wasi/core/src/swam/runtime/wasi/Core.scala [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+            F.pure(notcapable)
+          } else {
+            val stat =
+              if ((oflags & Oflags.Directory) == Oflags.Directory) {
+                // this is a directory, potentially to create
+                val linkOption =
+                  if ((dirflags & Lookupflags.SymlinkFollow) == 0)
+                    List(LinkOption.NOFOLLOW_LINKS)
+                  else
+                    Nil
```
[the remark this answers is not in the document]
Comment (satabin) [reply wasi/core/src/swam/runtime/wasi/Core.scala:265]: yeah I am not convinced about the readybility of `guard` :/

File: wasi/core/src/swam/runtime/wasi/Core.scala [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+                val stat = FdStat(Filetype.Directory, oflags, fsRightsBase, fsRightsInheriting, child, None)
+                if (Files.notExists(child, linkOption: _*) && (oflags & Oflags.Creat) == Oflags.Creat) {
+                  blocker
+                    .delay[F, Unit](Files.createDirectories(child))
```
[the remark this answers is not in the document]
Comment (satabin) [reply wasi/core/src/swam/runtime/wasi/Core.scala:269]: that's true, for the actual writing I might use it so why not here, that's true.

File: wasi/core/src/swam/runtime/wasi/Core.scala [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+  private def deallocateFd(fd: Fd): F[Unit] =
+    openFiles
+      .modify {
+        case files @ OpenFiles(descriptors, lastOpen, freed) =>
+          if (fd < 0 || fd >= lastOpen) {
+            (files, None)
+          } else {
+            val stat = descriptors(fd)
+            descriptors(fd) == null
```
[the remark this answers is not in the document]
Comment (satabin) [reply wasi/core/src/swam/runtime/wasi/Core.scala:339]: Silly me… Indeed.

File: build.sc (hunk 2 above)
Comment (satabin) [reply build.sc:45]: true.

**Comment by satabin** [comment]
@Jacarte FYI, this is my ongoing WASI core impementation.

**Comment by monperrus** [comment]
Wonderful!

**Pull request merged**